kernel/src/syscall/rt_sigtimedwait.rs

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::{sync::atomic::Ordering, time::Duration};
+
+use ostd::sync::Waiter;
+
+use super::SyscallReturn;
+use crate::{
+    prelude::*,
+    process::signal::{
+        constants::{SIGKILL, SIGSTOP},
+        sig_action::SigAction,
+        sig_mask::{SigMask, SigSet},
+        with_sigmask_changed, HandlePendingSignal,
+    },
+    time::{timespec_t, wait::ManagedTimeout},
+};
+
+pub fn sys_rt_sigtimedwait(
+    set_ptr: Vaddr,
+    info_ptr: Vaddr,
+    timeout_ptr: Vaddr,
+    sigset_size: usize,
+    ctx: &Context,
+) -> Result<SyscallReturn> {
+    debug!(
+        "set_ptr = 0x{:x}, info_ptr = 0x{:x}, timeout_ptr = 0x{:x}, sigset_size = {}",
+        set_ptr, info_ptr, timeout_ptr, sigset_size
+    );
+
+    // Validate sigset size
+    if sigset_size != size_of::<SigMask>() {
+        return_errno_with_message!(Errno::EINVAL, "invalid sigset size");
+    }
+
+    // Read the signal set
+    let mask = {
+        let mut set: SigSet = ctx.user_space().read_val(set_ptr)?;
+
+        // Remove SIGKILL and SIGSTOP as they cannot be waited for
+        set -= SIGKILL;
+        set -= SIGSTOP;
+
+        !set
+    };
+
+    // Read timeout if provided
+    let timeout = if timeout_ptr != 0 {
+        let timespec: timespec_t = ctx.user_space().read_val(timeout_ptr)?;
+        Some(Duration::try_from(timespec)?)
+    } else {
+        None
+    };
+
+    debug!(
+        "pid = {}, sig_mask = {:?}, timeout = {:?}",
+        ctx.process.pid(),
+        mask,
+        timeout
+    );
+
+    let block_list = ctx.posix_thread.sig_mask().load(Ordering::Relaxed);
+    with_sigmask_changed(
+        ctx,
+        |sig_mask| sig_mask & mask,
+        || {
+            // Fast path: If a signal is already pending, dequeue and return it immediately.
+            let sig_dispositions = ctx.process.sig_dispositions().lock();
+            let sig_disposition = sig_dispositions.lock();
+            while let Some(signal) = ctx.posix_thread.dequeue_signal(&mask) {
+                if info_ptr != 0 {
+                    let siginfo = signal.to_info();
+                    ctx.user_space().write_val(info_ptr, &siginfo)?;
+                }
+
+                // If the signal is ignored and not blocked, the signal will be directly dropped.
+                if !block_list.contains(signal.num())
+                    && sig_disposition.get(signal.num()) == SigAction::Ign
```

Comment:
You can use `sig_disposition.will_ignore(&signal)` instead of `sig_disposition.get(signal.num()) == SigAction::Ign` here. There are two cases when a signal will be ignored, the action is `SigAction::Ign`, or the action is `SigAction::Dfl`, and the default action is to ignore the signal.